File: java/src/main/behavioral/strategy/Context.java
package behavioral.strategy;

// Context: Maintains a reference to a strategy object
public class Context {
    private Strategy strategy;

    public Context(Strategy strategy) {
        this.strategy = strategy;
    }

    public void setStrategy(Strategy strategy) {
        this.strategy = strategy;
    }

    public int executeStrategy(int a, int b) {
        return strategy.execute(a, b);
    }
}

// Example Usage:
// public class Main {
//     public static void main(String[] args) {
//         Context context = new Context(new ConcreteStrategyAdd());
//
//         System.out.println("10 + 5 = " + context.executeStrategy(10, 5));
//
//         context.setStrategy(new ConcreteStrategySubtract());
//         System.out.println("10 - 5 = " + context.executeStrategy(10, 5));
//
//         context.setStrategy(new ConcreteStrategyMultiply());
//         System.out.println("10 * 5 = " + context.executeStrategy(10, 5));
//
//         // Expected Output:
//         // 10 + 5 = 15
//         // 10 - 5 = 5
//         // 10 * 5 = 50
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Using `if-else` Instead of Strategy Pattern:
//    - Mistake: Using multiple `if-else` statements instead of defining separate strategy classes.
//    - Fix: Use separate concrete strategy classes to follow the open-closed principle.
//
// 2. Allowing Context to Depend on Concrete Strategies:
//    - Mistake: Creating strategies inside the `Context` class.
//    - Fix: Inject strategies via constructor or setter methods.
//
// 3. Modifying Strategies Instead of Creating New Ones:
//    - Mistake: Mutating existing strategies instead of switching them.
//    - Fix: Use `setStrategy()` to swap strategy implementations at runtime.
//
// 4. Overusing the Strategy Pattern:
//    - Mistake: Applying Strategy when a simple function reference would be enough.
//    - Fix: Use Strategy Pattern only when multiple interchangeable behaviors are required.

File: python/src/behavioral/strategy/context.py
from .strategy import Strategy


class Context:
    """
    The Context maintains a reference to a Strategy object and delegates
    the work to it instead of implementing multiple algorithm variations.
    """

    def __init__(self, strategy: Strategy):
        self._strategy: Strategy = strategy

    def set_strategy(self, strategy: Strategy) -> None:
        self._strategy = strategy

    def execute_strategy(self, a: int, b: int) -> int:
        return self._strategy.execute(a, b)


# Example Usage:
# from context import Context
# from concrete_strategy_a import ConcreteStrategyA
# from concrete_strategy_b import ConcreteStrategyB
#
# context = Context(ConcreteStrategyA())
# print(context.execute_strategy(5, 3))
# # Output: 8
#
# context.set_strategy(ConcreteStrategyB())
# print(context.execute_strategy(5, 3))
# # Output: 15
#
#
# Common Mistakes to Avoid:
#
# 1. Adding conditional logic inside the Context instead of strategies.
#    - Fix: Move algorithm variations into separate Strategy classes.
#
# 2. Creating too many strategies for trivial operations.
#    - Fix: Only use Strategy when you truly need interchangeable algorithms.
#
# 3. Forgetting to make Strategy methods consistent.
#    - Fix: Ensure all strategies implement the same method signature.
